chain/client/src/chunks_delay_tracker.rs

```diff
@@ -0,0 +1,188 @@
+use std::collections::{BTreeMap, HashMap};
+use std::time::Instant;
+
+use near_primitives::types::{BlockHeight, ShardId};
+
+use crate::metrics;
+use std::cmp::max;
+
+/// Provides monitoring information about the delays of receiving blocks and their corresponding chunks.
+/// Keeps timestamps of a limited number of blocks before the current head.
+/// Keeps timestamps of all blocks and chunks from the future.
+/// Doesn't make any assumptions about shard layout or the number of shards. If the head was moved forward, we assume that all requirements were satisfied.
+/// This object provides two metrics:
+/// 1) max delay between discovering a block and receiving the last chunk from that block.
+/// 2) number of blocks with missing chunks.
+/// If a chunk or a block is received multiple times, only the first time is recorded.
+#[derive(Debug, Default)]
+pub(crate) struct ChunksDelayTracker {
+    heights: BTreeMap<BlockHeight, HeightInfo>,
+    // Used as an optimization to delete old blocks only when the head is updated.
+    prev_head_height: BlockHeight,
+}
+
+const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
+
+#[derive(Debug, Default)]
+struct HeightInfo {
+    block_received: Option<Instant>,
+    chunks_received: HashMap<ShardId, Instant>,
+}
+
+impl ChunksDelayTracker {
+    // Blocks below this heights are not tracked.
+    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
+        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    }
+
+    fn remove_old_entries(&mut self, head_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> = self
+            .heights
+            .range(..ChunksDelayTracker::lowest_height(head_height))
+            .map(|(&k, _v)| k)
+            .collect();
+        for h in heights_to_drop {
+            self.heights.remove(&h);
+        }
+    }
+
+    fn update_chunks_metric(&mut self, head_height: BlockHeight) {
+        if head_height == self.prev_head_height {
+            return;
+        }
+        let mut max_delay = 0;
+        for (_, v) in self.heights.range(..=head_height) {
+            if let Some(block_received) = v.block_received {
+                if let Some(latest_chunk) = v.chunks_received.values().max() {
+                    if latest_chunk > &block_received {
+                        let delay = latest_chunk.duration_since(block_received).as_micros();
+                        max_delay = max(max_delay, delay);
+                    }
+                }
+            }
+        }
+        near_metrics::set_gauge(&metrics::CHUNKS_RECEIVING_DELAY_US, max_delay as i64);
+        self.prev_head_height = head_height;
+    }
+
+    fn update_blocks_missing_chunks_metric(&mut self, head_height: BlockHeight) {
+        let value = if let Some((&latest, _)) = self.heights.iter().next_back() {
+            latest.saturating_sub(head_height)
+        } else {
+            0
+        };
+        near_metrics::set_gauge(&metrics::BLOCKS_AHEAD_OF_HEAD, value as i64);
```

Comment:
I think this metric is not that different from the block difference metric we already have right now? Also, it's not necessarily true that a block is received but not processed always has missing chunks.

Reply:
If I understand correctly, the block difference metric is computed as the difference of `max(head)` across all nodes and `head` of the current node.
`BLOCKS_AHEAD_OF_HEAD` metric tries to do that on the node itself.

> it's not necessarily true that a block is received but not processed always has missing chunks.
Renamed the function and cleaned up a few comments. Thanks.